File: inst/tinytest/test-analytic.R
source("helpers.R")
using("marginaleffects")

# lm quadratic
set.seed(1027)
f <- y ~ x + I(x^2)
truth <- function(x) 1 + 2 * x
N <- 100000
dat <- data.frame(x = rnorm(N))
dat$y <- 1 + 1 * dat$x + 1 * dat$x^2 + rnorm(N)
mod <- lm(f, dat)
nd <- datagrid(newdata = dat, x = c(-2:2))
res <- slopes(mod, newdata = nd)
res$truth <- truth(res$x)
expect_equivalent(res$estimate, res$truth, tolerance = .01)


# lm log
set.seed(30)
f <- y ~ log(x)
truth <- function(x) 1 / x
N <- 10000
dat <- data.frame(x = runif(N))
dat$y <- log(dat$x) + rnorm(N)
mod <- lm(f, dat)
nd <- datagrid(newdata = dat, x = c(1:4))
res <- slopes(mod, newdata = nd)
res$truth <- truth(res$x)
expect_equivalent(res$estimate, res$truth, tolerance = .01)



# logit
set.seed(2000)
f <- y ~ x
beta0 <- 1
beta1 <- .2
truth <- function(x) beta1 * dlogis(beta0 + beta1 * x)
N <- 1e5
dat <- data.frame(x = rnorm(N, sd = 3))
dat$y <- rbinom(N, 1, pr = plogis(beta0 + beta1 * dat$x))
mod <- glm(f, data = dat, family = binomial)
nd <- datagrid(newdata = dat, x = c(-10:10))
res <- slopes(mod, newdata = nd)
res$truth <- truth(res$x)
expect_equivalent(res$estimate, res$truth, tolerance = .01)





############################################################################
#  golder tests copied from the `margins` github repository on 2021-09-18  #
#  LICENSE: MIT Thomas J. Leeper                                           #
############################################################################

# tests based on formulae from Matt Golder's OLS examples, for numerical accuracy and precision

# example data for tests
set.seed(1)
n <- 25L
d <- data.frame(w = rnorm(n),
        x = rnorm(n),
        z = rnorm(n))
d[["y"]] <- with(d, w + x + z + w*x + w*z + x*z * w*x*z + rnorm(n))

# set comparison tolerance
tol <- 0.001
tol_se <- 0.005

# http://mattgolder.com/wp-content/uploads/2015/05/standarderrors1.png

# Golder Interaction Case 1a/1b correct
f1.1 <- y ~ x + z + x:z
m <- lm(f1.1, data = d)
marg <- slopes(m)
# ME with respect to x
dydx <- coef(m)["x"] + (d$z * coef(m)["x:z"])
sedydx <- sqrt(vcov(m)["x","x"] + (d$z^2 * vcov(m)["x:z","x:z"]) + (2 * d$z * vcov(m)["x","x:z"]))
expect_equivalent(as.numeric(marg$estimate[marg$term == "x"]), dydx, tolerance = tol, label = "dy/dx correct")
expect_equivalent(sedydx, as.numeric(marg$std.error[marg$term == "x"]), tolerance = tol_se, label = "Var(dy/dx) correct")
# ME with respect to z
dydz <- coef(m)["z"] + (d$x * coef(m)["x:z"])
sedydz <- sqrt(vcov(m)["z","z"] + (d$x^2 * vcov(m)["x:z","x:z"]) + (2 * d$x * vcov(m)["z","x:z"]))
expect_equivalent(as.numeric(marg$estimate[marg$term == "z"]), dydz, tolerance = tol, label = "dy/dz correct")
expect_equivalent(sedydz, as.numeric(marg$std.error[marg$term == "z"]), tolerance = tol_se, label = "Var(dy/dz) correct")


# Golder Interaction Case 2 correct
f1.2 <- y ~ x + z + w + x:z + z:w
m <- lm(f1.2, data = d)
marg <- slopes(m)
dydx <- coef(m)["x"] + (d$z * coef(m)["x:z"])
sedydx <- sqrt(vcov(m)["x","x"] + (d$z^2 * vcov(m)["x:z","x:z"]) + (2 * d$z * vcov(m)["x","x:z"]))
expect_equivalent(as.numeric(marg$estimate[marg$term == "x"]), dydx, tolerance = tol, label = "dy/dx correct")
expect_equivalent(sedydx, as.numeric(marg$std.error[marg$term == "x"]), tolerance = tol_se, label = "Var(dy/dx) correct")


# Golder Interaction Case 3 correct
f1.3 <- y ~ x + z + w + x:z + x:w + z:w
m <- lm(f1.3, data = d)
marg <- slopes(m)
dydx <- coef(m)["x"] + (d$z * coef(m)["x:z"]) + (d$w * coef(m)["x:w"])
sedydx <- sqrt(vcov(m)["x","x"] + (d$z^2 * vcov(m)["x:z","x:z"]) + (d$w^2 * vcov(m)["x:w","x:w"]) + 
           (2 * d$z * vcov(m)["x","x:z"]) + (2 * d$w * vcov(m)["x","x:w"]) + (2 * d$z * d$w * vcov(m)["x:z","x:w"]) )
expect_equivalent(as.numeric(marg$estimate[marg$term == "x"]), dydx, tolerance = tol, label = "dy/dx correct")
expect_equivalent(sedydx, as.numeric(marg$std.error[marg$term == "x"]), tolerance = tol_se, label = "Var(dy/dx) correct")


# Golder Interaction Case 4 correct
f1.4 <- y ~ x + z + w + x:z + x:w + z:w + x:z:w
m <- lm(f1.4, data = d)
marg <- slopes(m)
dydx <- coef(m)["x"] + (d$z * coef(m)["x:z"]) + (d$w * coef(m)["x:w"]) + (d$z * d$w * coef(m)["x:z:w"])
sedydx <- sqrt(vcov(m)["x","x"] + (d$z^2 * vcov(m)["x:z","x:z"]) + (d$w^2 * vcov(m)["x:w","x:w"]) + 
           (d$z^2 * d$w^2 * vcov(m)["x:z:w","x:z:w"]) + (2 * d$z * vcov(m)["x","x:z"]) + 
           (2 * d$w * vcov(m)["x","x:w"]) + (2 * d$z * d$w * vcov(m)["x","x:z:w"]) + 
           (2 * d$z * d$w * vcov(m)["x:z","x:w"]) + (2 * d$w * d$z^2 * vcov(m)["x:z","x:z:w"]) +
           (2 * d$z * d$w^2 * vcov(m)["x:w","x:z:w"]) )
expect_equivalent(as.numeric(marg$estimate[marg$term == "x"]), dydx, tolerance = tol, label = "dy/dx correct")
expect_equivalent(sedydx, as.numeric(marg$std.error[marg$term == "x"]), tolerance = tol_se, label = "Var(dy/dx) correct")


# Golder Quadratic Case 1 correct
f2.1 <- y ~ x + I(x^2)
m <- lm(f2.1, data = d)
marg <- slopes(m)
dydx <- coef(m)["x"] + (2 * coef(m)["I(x^2)"] * d$x)
sedydx <- sqrt(vcov(m)["x","x"] + (4 * d$x^2 * vcov(m)["I(x^2)","I(x^2)"]) + (4 * d$x * vcov(m)["x","I(x^2)"]))
expect_equivalent(as.numeric(marg$estimate[marg$term == "x"]), dydx, tolerance = tol, label = "dy/dx correct")
expect_equivalent(sedydx, as.numeric(marg$std.error[marg$term == "x"]), tolerance = tol_se, label = "Var(dy/dx) correct")


# Golder Quadratic Case 2 correct
f2.2 <- y ~ x + I(x^2) + z
m <- lm(f2.2, data = d)
marg <- slopes(m)
dydx <- coef(m)["x"] + (2 * coef(m)["I(x^2)"] * d$x)
sedydx <- sqrt(vcov(m)["x","x"] + (4 * d$x^2 * vcov(m)["I(x^2)","I(x^2)"]) + (4 * d$x * vcov(m)["x","I(x^2)"]))
expect_equivalent(as.numeric(marg$estimate[marg$term == "x"]), dydx, tolerance = tol, label = "dy/dx correct")
expect_equivalent(sedydx, as.numeric(marg$std.error[marg$term == "x"]), tolerance = tol_se, label = "Var(dy/dx) correct")


# Golder Quadratic Case 3a/3b correct
f2.3 <- y ~ x + I(x^2) + z + x:z
m <- lm(f2.3, data = d)
marg <- slopes(m)
# ME with respect to x
dydx <- coef(m)["x"] + (2 * coef(m)["I(x^2)"] * d$x) + (d$z * coef(m)["x:z"])
sedydx <- sqrt(vcov(m)["x","x"] + (4 * d$x^2 * vcov(m)["I(x^2)","I(x^2)"]) + (d$z^2 * vcov(m)["x:z","x:z"]) +
           (4 * d$x * vcov(m)["x","I(x^2)"]) + (2 * d$z * vcov(m)["x","x:z"]) + (4 * d$x * d$z * vcov(m)["I(x^2)", "x:z"]) )
expect_equivalent(as.numeric(marg$estimate[marg$term == "x"]), dydx, tolerance = tol, label = "dy/dx correct")
expect_equivalent(sedydx, as.numeric(marg$std.error[marg$term == "x"]), tolerance = tol_se, label = "Var(dy/dx) correct")
# ME with respect to z
dydz <- coef(m)["z"] + (d$x * coef(m)["x:z"])
sedydz <- sqrt(vcov(m)["z","z"] + (d$x^2 * vcov(m)["x:z","x:z"]) + (2 * d$x * vcov(m)["z","x:z"]))
expect_equivalent(as.numeric(marg$estimate[marg$term == "z"]), dydz, tolerance = tol, label = "dy/dz correct")
expect_equivalent(sedydz, as.numeric(marg$std.error[marg$term == "z"]), tolerance = tol_se, label = "Var(dy/dz) correct")


# Golder Quadratic Case 4a/4b correct
f2.4 <- y ~ x + I(x^2) + z + x:z + I(x^2):z
m <- lm(f2.4, data = d)
marg <- slopes(m)
# ME with respect to x
dydx <- coef(m)["x"] + (2 * coef(m)["I(x^2)"] * d$x) + (d$z * coef(m)["x:z"]) + (2 * d$x * d$z * coef(m)["I(x^2):z"])
sedydx <- sqrt( vcov(m)["x","x"] + (4 * d$x^2 * vcov(m)["I(x^2)","I(x^2)"]) + (d$z^2 * vcov(m)["x:z","x:z"]) +
           (4 * (d$x^2) * (d$z^2) * vcov(m)["I(x^2):z","I(x^2):z"]) + 
           (4 * d$x * vcov(m)["x","I(x^2)"]) + (2 * d$z * vcov(m)["x","x:z"]) + 
           (4 * d$x * d$z * vcov(m)["I(x^2)", "x:z"]) +
           (4 * d$x * d$z * vcov(m)["x","I(x^2):z"]) + 
           (8 * (d$x^2) * d$z * vcov(m)["I(x^2)","I(x^2):z"]) + 
           (4 * d$x * (d$z^2) * vcov(m)["x:z","I(x^2):z"]) )
expect_equivalent(as.numeric(marg$estimate[marg$term == "x"]), dydx, tolerance = tol, label = "dy/dx correct")
expect_equivalent(sedydx, as.numeric(marg$std.error[marg$term == "x"]), tolerance = tol_se, label = "Var(dy/dx) correct")
# ME with respect to z
dydz <- coef(m)["z"] + (d$x * coef(m)["x:z"]) + (d$x^2 * coef(m)["I(x^2):z"])
sedydz <- sqrt(vcov(m)["z","z"] + (d$x^2 * vcov(m)["x:z","x:z"]) + (d$x^4 * vcov(m)["I(x^2):z","I(x^2):z"]) + 
           (2 * d$x * vcov(m)["z","x:z"]) + (2 * (d$x^2) * vcov(m)["z","I(x^2):z"]) + 
           (2 * (d$x^3) * vcov(m)["x:z","I(x^2):z"]) )
expect_equivalent(as.numeric(marg$estimate[marg$term == "z"]), dydz, tolerance = tol, label = "dy/dz correct")
expect_equivalent(sedydz, as.numeric(marg$std.error[marg$term == "z"]), tolerance = tol_se, label = "Var(dy/dz) correct")

rm(list = ls())

File: python/tests/test_analytic.py
import numpy as np
import polars as pl
import statsmodels.formula.api as smf
import statsmodels.api as sm
from scipy.stats import logistic
from polars.testing import assert_series_equal

from marginaleffects import *


# Translated from https://github.com/vincentarelbundock/marginaleffects/blob/main/inst/tinytest/test-analytic.R


def test_linear_quadratic():
    np.random.seed(1)
    f = "y ~ x + I(x**2)"

    def truth(x):
        return 1 + 2 * x

    N = 100000
    dat = pl.DataFrame({"x": np.random.normal(size=N)})
    dat = dat.with_columns(
        y=(1 + pl.col("x") + pl.col("x") ** 2 + np.random.normal(size=N))
    )
    mod = smf.ols(f, dat.to_pandas()).fit()
    nd = datagrid(newdata=dat, x=range(-2, 3))
    res = slopes(mod, newdata=nd)
    res = res.with_columns(truth=truth(pl.col("x")).cast(pl.Float64))
    assert_series_equal(res["estimate"], res["truth"], check_names=False, abs_tol=0.01)


def test_linear_log():
    np.random.seed(1)
    f = "y ~ np.log(x)"

    def truth(x):
        return 1 / x

    N = 10000
    dat = pl.DataFrame({"x": np.random.uniform(size=N)})
    dat = dat.with_columns(y=np.log(pl.col("x")) + np.random.normal(size=N))
    mod = smf.ols(f, dat.to_pandas()).fit()
    nd = datagrid(newdata=dat, x=range(1, 5))
    res = slopes(mod, newdata=nd)
    res = res.with_columns(truth=truth(pl.col("x")).cast(pl.Float64))
    assert_series_equal(res["estimate"], res["truth"], check_names=False, abs_tol=0.02)


def test_logit():
    np.random.seed(1)
    f = "y ~ x"
    beta0 = 1
    beta1 = 0.2

    def truth(x):
        return beta1 * logistic.pdf(beta0 + beta1 * x)

    N = int(1e5)
    dat = pl.DataFrame({"x": np.random.normal(size=N)})
    pr = logistic.cdf(beta0 + beta1 * dat["x"])
    dat = dat.with_columns(y=np.random.binomial(1, pr, size=N))
    mod = smf.glm(f, data=dat, family=sm.families.Binomial()).fit()
    nd = datagrid(newdata=dat, x=range(-10, 11))
    res = slopes(mod, newdata=nd)
    tru = truth(res["x"])
    res = res.with_columns(truth=tru)
    assert_series_equal(res["estimate"], res["truth"], check_names=False, abs_tol=0.01)


############################################################################
#  golder tests copied to the R version of `marginaleffects`               #
#  from the `margins` github repository on 2021-09-18                      #
#  and translated to Python on 2024-02-15                                  #
#  LICENSE: MIT Thomas J. Leeper                                           #
############################################################################

# tests based on formulae from Matt Golder's OLS examples, for numerical accuracy and precision
# http://mattgolder.com/wp-content/uploads/2015/05/standarderrors1.png


def test_golder():
    # example data for tests
    np.random.seed(1)
    N = 25
    dat = pl.DataFrame(
        {
            "w": np.random.normal(size=N),
            "x": np.random.normal(size=N),
            "z": np.random.normal(size=N),
        }
    )
    dat = dat.with_columns(
        y=pl.col("w")
        + pl.col("x")
        + pl.col("z")
        + pl.col("w") * pl.col("x")
        + pl.col("w") * pl.col("z")
        + pl.col("x") * pl.col("z")
        + pl.col("w") * pl.col("x") * pl.col("z")
        + np.random.normal(size=N)
    )

    # set comparison tolerance
    tol = 0.001
    tol_se = 0.01

    # Golder Interaction Case 1a/1b correct
    f1_1 = "y ~ x + z + x:z"
    mod = smf.ols(f1_1, data=dat).fit()
    res = slopes(mod)
    dydx = mod.params["x"] + (dat["z"] * mod.params["x:z"])
    vcov = mod.cov_params()
    sedydx = np.sqrt(
        vcov.at["x", "x"]
        + dat["z"] ** 2 * vcov.at["x:z", "x:z"]
        + 2 * dat["z"] * vcov.at["x", "x:z"]
    )
    assert_series_equal(
        res.filter(pl.col("term") == "x")["estimate"],
        dydx,
        check_names=False,
        abs_tol=tol,
    )
    assert_series_equal(
        res.filter(pl.col("term") == "x")["std_error"],
        sedydx,
        check_names=False,
        abs_tol=tol_se,
    )

    # Golder Interaction Case 2 correct
    f1_2 = "y ~ x + z + w + x:z + z:w"
    mod = smf.ols(f1_2, data=dat).fit()
    res = slopes(mod)
    dydx = mod.params["x"] + (dat["z"] * mod.params["x:z"])
    vcov = mod.cov_params()
    sedydx = np.sqrt(
        vcov.at["x", "x"]
        + dat["z"] ** 2 * vcov.at["x:z", "x:z"]
        + 2 * dat["z"] * vcov.at["x", "x:z"]
    )
    assert_series_equal(
        res.filter(pl.col("term") == "x")["estimate"],
        dydx,
        check_names=False,
        abs_tol=tol,
    )
    assert_series_equal(
        res.filter(pl.col("term") == "x")["std_error"],
        sedydx,
        check_names=False,
        abs_tol=tol_se,
    )

    # Golder Interaction Case 3 correct
    f1_3 = "y ~ x + z + w + x:z + x:w + z:w + x:z:w"
    mod = smf.ols(f1_3, data=dat).fit()
    res = slopes(mod)
    dydx = (
        mod.params["x"]
        + dat["z"] * mod.params["x:z"]
        + dat["w"] * mod.params["x:w"]
        + dat["z"] * dat["w"] * mod.params["x:z:w"]
    )
    vcov = mod.cov_params()
    sedydx = np.sqrt(
        vcov.at["x", "x"]
        + dat["z"] ** 2 * vcov.at["x:z", "x:z"]
        + dat["w"] ** 2 * vcov.at["x:w", "x:w"]
        + dat["z"] ** 2 * dat["w"] ** 2 * vcov.at["x:z:w", "x:z:w"]
        + 2 * dat["z"] * vcov.at["x", "x:z"]
        + 2 * dat["w"] * vcov.at["x", "x:w"]
        + 2 * dat["z"] * dat["w"] * vcov.at["x", "x:z:w"]
        + 2 * dat["z"] * dat["w"] * vcov.at["x:z", "x:w"]
        + 2 * dat["z"] ** 2 * dat["w"] * vcov.at["x:z", "x:z:w"]
        + 2 * dat["z"] * dat["w"] ** 2 * vcov.at["x:w", "x:z:w"]
    )
    assert_series_equal(
        res.filter(pl.col("term") == "x")["estimate"],
        dydx,
        check_names=False,
        abs_tol=tol,
    )
    assert_series_equal(
        res.filter(pl.col("term") == "x")["std_error"],
        sedydx,
        check_names=False,
        abs_tol=tol_se,
    )

    # Golder Quadratic Case 1 correct
    f2_1 = "y ~ x + I(x**2)"
    mod = smf.ols(f2_1, data=dat.to_pandas()).fit()
    res = slopes(mod)
    dydx = mod.params["x"] + (2 * mod.params["I(x ** 2)"] * dat["x"])
    vcov = mod.cov_params()
    sedydx = np.sqrt(
        vcov.at["x", "x"]
        + 4 * dat["x"] ** 2 * vcov.at["I(x ** 2)", "I(x ** 2)"]
        + 4 * dat["x"] * vcov.at["x", "I(x ** 2)"]
    )
    assert_series_equal(
        res.filter(pl.col("term") == "x")["estimate"],
        dydx,
        check_names=False,
        abs_tol=tol,
    )
    assert_series_equal(
        res.filter(pl.col("term") == "x")["std_error"],
        sedydx,
        check_names=False,
        abs_tol=tol_se,
    )

    # Golder Quadratic Case 2 correct
    f2_2 = "y ~ x + I(x**2) + z"
    mod = smf.ols(f2_2, data=dat.to_pandas()).fit()
    res = slopes(mod)
    dydx = mod.params["x"] + (2 * mod.params["I(x ** 2)"] * dat["x"])
    vcov = mod.cov_params()
    sedydx = np.sqrt(
        vcov.at["x", "x"]
        + 4 * dat["x"] ** 2 * vcov.at["I(x ** 2)", "I(x ** 2)"]
        + 4 * dat["x"] * vcov.at["x", "I(x ** 2)"]
    )
    assert_series_equal(
        res.filter(pl.col("term") == "x")["estimate"],
        dydx,
        check_names=False,
        abs_tol=tol,
    )
    assert_series_equal(
        res.filter(pl.col("term") == "x")["std_error"],
        sedydx,
        check_names=False,
        abs_tol=tol_se,
    )

    # Golder Quadratic Case 3a/3b correct
    f2_3 = "y ~ x + I(x**2) + z + x:z"
    mod = smf.ols(f2_3, data=dat.to_pandas()).fit()
    res = slopes(mod)
    vcov = mod.cov_params()

    # ME with respect to x
    dydx = (
        mod.params["x"]
        + 2 * mod.params["I(x ** 2)"] * dat["x"]
        + dat["z"] * mod.params["x:z"]
    )
    sedydx = np.sqrt(
        vcov.at["x", "x"]
        + 4 * dat["x"] ** 2 * vcov.at["I(x ** 2)", "I(x ** 2)"]
        + dat["z"] ** 2 * vcov.at["x:z", "x:z"]
        + 4 * dat["x"] * vcov.at["x", "I(x ** 2)"]
        + 2 * dat["z"] * vcov.at["x", "x:z"]
        + 4 * dat["x"] * dat["z"] * vcov.at["I(x ** 2)", "x:z"]
    )
    assert_series_equal(
        res.filter(pl.col("term") == "x")["estimate"],
        dydx,
        check_names=False,
        abs_tol=tol,
    )
    assert_series_equal(
        res.filter(pl.col("term") == "x")["std_error"],
        sedydx,
        check_names=False,
        abs_tol=tol_se,
    )

    # ME with respect to z
    dydz = mod.params["z"] + (dat["x"] * mod.params["x:z"])
    sedydz = np.sqrt(
        vcov.at["z", "z"]
        + dat["x"] ** 2 * vcov.at["x:z", "x:z"]
        + 2 * dat["x"] * vcov.at["z", "x:z"]
    )
    assert_series_equal(
        res.filter(pl.col("term") == "z")["estimate"],
        dydz,
        check_names=False,
        abs_tol=tol,
    )
    assert_series_equal(
        res.filter(pl.col("term") == "z")["std_error"],
        sedydz,
        check_names=False,
        abs_tol=tol_se,
    )

    # Golder Quadratic Case 4a/4b correct
    f2_4 = "y ~ x + I(x**2) + z + x:z + I(x**2):z"
    mod = smf.ols(f2_4, data=dat.to_pandas()).fit()
    res = slopes(mod)
    vcov = mod.cov_params()

    # ME with respect to x
    dydx = (
        mod.params["x"]
        + (2 * mod.params["I(x ** 2)"] * dat["x"])
        + (dat["z"] * mod.params["x:z"])
        + (2 * dat["x"] * dat["z"] * mod.params["I(x ** 2):z"])
    )
    sedydx = np.sqrt(
        vcov.at["x", "x"]
        + 4 * dat["x"] ** 2 * vcov.at["I(x ** 2)", "I(x ** 2)"]
        + dat["z"] ** 2 * vcov.at["x:z", "x:z"]
        + 4 * dat["x"] ** 2 * dat["z"] ** 2 * vcov.at["I(x ** 2):z", "I(x ** 2):z"]
        + 4 * dat["x"] * vcov.at["x", "I(x ** 2)"]
        + 2 * dat["z"] * vcov.at["x", "x:z"]
        + 4 * dat["x"] * dat["z"] * vcov.at["I(x ** 2)", "x:z"]
        + 4
        * dat["x"]
        * dat["z"]
        * vcov.at[
            "x",
            "I(x ** 2):z",
        ]
        + 8 * dat["x"] ** 2 * dat["z"] * vcov.at["I(x ** 2)", "I(x ** 2):z"]
        + 4 * dat["x"] * dat["z"] ** 2 * vcov.at["x:z", "I(x ** 2):z"]
    )
    assert_series_equal(
        res.filter(pl.col("term") == "x")["estimate"],
        dydx,
        check_names=False,
        abs_tol=tol,
    )
    assert_series_equal(
        res.filter(pl.col("term") == "x")["std_error"],
        sedydx,
        check_names=False,
        abs_tol=tol_se,
    )

    # ME with respect to z
    dydz = (
        mod.params["z"]
        + dat["x"] * mod.params["x:z"]
        + dat["x"] ** 2 * mod.params["I(x ** 2):z"]
    )
    sedydz = np.sqrt(
        vcov.at["z", "z"]
        + dat["x"] ** 2 * vcov.at["x:z", "x:z"]
        + dat["x"] ** 4 * vcov.at["I(x ** 2):z", "I(x ** 2):z"]
        + 2 * dat["x"] * vcov.at["z", "x:z"]
        + 2 * dat["x"] ** 2 * vcov.at["z", "I(x ** 2):z"]
        + 2 * dat["x"] ** 3 * vcov.at["x:z", "I(x ** 2):z"]
    )
    assert_series_equal(
        res.filter(pl.col("term") == "z")["estimate"],
        dydz,
        check_names=False,
        abs_tol=tol,
    )
    assert_series_equal(
        res.filter(pl.col("term") == "z")["std_error"],
        sedydz,
        check_names=False,
        abs_tol=tol_se,
    )
